<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><!-- InstanceBegin template="/Templates/ChromaseqTemplate.dwt" codeOutsideHTMLIsLocked="false" -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />


<!-- InstanceBeginEditable name="doctitle" -->
<title>Submitting sequences to GenBank</title>
<!-- InstanceEndEditable -->

<![endif]--><!--[if IE]>
<style type="text/css"> 
/* place css fixes for all versions of IE in this conditional comment */
.thrColFixHdr #sidebar2, .thrColFixHdr #sidebar1 { padding-top: 30px; }
.thrColFixHdr #mainContent { zoom: 1; }
/* the above proprietary zoom property gives IE the hasLayout it needs to avoid several bugs */
</style>
<![endif]-->

<style type="text/css">
<!--
-->
</style>
<link href="CSS/mesquitePackage.css" rel="stylesheet" type="text/css" />
<link href="SpryAssets/SpryCollapsiblePanel.css" rel="stylesheet" type="text/css" />

<script src="SpryAssets/SpryCollapsiblePanel.js" type="text/javascript"></script>
<script type="text/javascript" src="SpryAssets/SpryURLUtils.js"></script>
<script type="text/javascript" src="javascript/mesquiteChromaseq.js"></script>

<!-- InstanceBeginEditable name="head" --><!-- InstanceEndEditable -->
</head>

<body bgcolor="#f4f0e2" class="thrColFixHdr" >

<div id="container">
  <div id="header">
    <table width="100%" border="0" cellpadding="4">
  <tr>
    <td width="51%"><strong id="PackageTitle">Chromaseq</strong></td>
    <td width="33%" align="right"><div id="subtitle"><!-- InstanceBeginEditable name="subtitle" -->Submitting
      Sequences to GenBank<!-- InstanceEndEditable --></div></td>
    <td width="2%">&nbsp;</td>
    <td width="14%"><p>Version 1.5 <br />
      27 December
2018<br />
      </p></td>
  </tr>
</table>

</div>
  
  
<div id="sidebar1">
  
  
  
  
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('index.html')">Home</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('overview.html')">Overview</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('download.html')">Download</a>
  </div>

  <div id="GettingStartedPanel" class="CollapsiblePanel">
    <div class="CollapsiblePanelTab" tabindex="0">Getting Started ˇ</div>
    <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('download.html');">Download</a></td></tr>
  			<tr><td><a href="javascript:pageLink('installation.html');">Installation</a></td></tr>
  			<tr><td><a href="javascript:pageLink('quickstart.html');">Quick Start</a></td></tr>
  			<tr><td><a href="javascript:pageLink('tutorial.html');">Tutorial</a></td></tr>
		</table>
    </div>
  </div>
  <div id="PreparingDataPanel" class="CollapsiblePanel">
    <div class="CollapsiblePanelTab" tabindex="0">Preparing Data ˇ</div>
    <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('chromatograms.html');">Chromatograms</a></td></tr>
  			<tr><td><a href="javascript:pageLink('fileNamingRules.html');">File Naming Rules</a></td></tr>
  			<tr><td><a href="javascript:pageLink('sampleCode.html');">Samples & Sequence Names</a></td></tr>
  			<tr><td><a href="javascript:pageLink('primers.html');">Primer & Gene Information</a></td></tr>
		</table>
    </div>
  </div>
  
<div id="ProcessingChromatoPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Processing Chromatograms ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr> <td><a href="javascript:pageLink('preparation.html');">Preparation & Processing</a></td></tr>
  			<tr><td><a href="javascript:pageLink('phredPhrap.html');">Phred & Phrap Locations & Options</a></td></tr>
  			<tr><td><a href="javascript:pageLink('postPhrap.html');">Post-Phrap Options</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('quality.html');">Quality Scores & Colors</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('segregatingChromatograms.html');">Segregating Chromatogram Files</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('appendingSequences.html');">Processing Additional Chromatograms</a></td></tr>
		</table>
  </div>
</div>


<div id="ChromatogramViewerPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Chromatogram Viewer ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('viewer.html');">Viewing Chromatograms</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('callingBases.html');">Altering Base Calls</a></td></tr>
  			<tr><td><a href="javascript:pageLink('conflicts.html');">Detecting Conflicts</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('primerSites.html');">Displaying Primer Sites</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('extraSequences.html');">Multiple Reads/Contigs at Once</a></td></tr>
  			<tr><td><a href="javascript:pageLink('display.html');">Other Display Options</a></td></tr>
			<tr><td><a href="javascript:pageLink('navigation.html');">Navigation</a></td></tr>
		</table>
  </div>
</div>

<div id="GenBankPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">GenBank Submission ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('genbank.html');">GenBank via tbl2asn</a></td></tr>
  			<tr><td><a href="javascript:pageLink('genbankSequin.html');">GenBank via Sequin</a></td></tr>
		</table>
  </div>
</div>

<div id="OtherFeaturesPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Other Features ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('listWindow.html');">List Window Features</a></td></tr>
  			<tr><td><a href="javascript:pageLink('matrixCells.html');">Matrix Colors</a></td></tr>
  			<tr><td><a href="javascript:pageLink('purging.html');">Purging Chromaseq Data</a></td></tr>
  			<tr><td><a href="javascript:pageLink('helpfulMesquite.html');">Helpful Mesquite Features</a></td></tr>
		</table>
  </div>
</div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('workflow.html')">Workflow Suggestions</a>
  </div>

<div id="HelpPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Help & Info ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('tutorial.html');">Tutorial</a></td></tr>
  			<tr><td><a href="javascript:pageLink('faq.html');">FAQ</a></td></tr>
  			<tr><td><a href="javascript:pageLink('troubleshooting.html');">Troubleshooting</a></td></tr>
  			<tr><td><a href="javascript:pageLink('chromaseqList.html');">Chromaseq E-mail List</a></td></tr>
  			<tr><td><a href="javascript:pageLink('limitations.html');">Limitations</a></td></tr>
  			<tr><td><a href="javascript:pageLink('otherPrograms.html');">Chromaseq &amp; Other Programs</a></td></tr>
  			<tr><td><a href="javascript:pageLink('future.html');">Future Directions</a></td></tr>
	</table>
  </div>
</div>

<div id="TechnicalPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Technical Details ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('sourceCode.html');">Source Code</a></td></tr>
  			<tr><td><a href="javascript:pageLink('technicalFAQ.html');">Technical FAQ</a></td></tr>
		</table>
  </div>
</div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('knownIssues.html')">Known Issues</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('versions.html')">Versions</a>
  </div>

<div id="CreditPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Citation & Credit ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('citation.html');">How to Cite</a></td></tr>
  			<tr><td><a href="javascript:pageLink('acknowledgements.html');">Acknowledgements</a></td></tr>
		</table>
  </div>
</div>


<h3>&nbsp;</h3>
  <!-- end #sidebar1 --></div>

  <div id="mainContent">
    
<!-- InstanceBeginEditable name="content" -->
<h2>Preparation of files for submission using tbl2asn </h2>
<p>Chromaseq has two features designed to aid submitting sequences contained in a Mesquite file to GenBank. The primary tool, described here, allows for preparing files for use by tbl2asn. There is also a <a href="genbankSequin.html">legacy tool</a> for submitting sequences via Sequin. </p>
<p>This system is built to submit sequences from one matrix at a time.</p>
<p>To prepare
  files for submission by tbl2asn, you will need the following: </p>
<ol>
  <li>The <a href="#downloadTbl2asn">tbl2asn program</a> from NCBI.</li>
  <li>A <a href="#submissionTemplate">submission template file</a> containing information about your paper.</li>
  <li>A <a href="#sampleTable">tab-delimited text file</a> containing information about each sample whose data will be submitted. This file contains organism's name, authority, locality data,
    etc. This is called the &quot;OTU ID code Database&quot; or &quot;OTU ID code DB&quot; file in Mesquite.</li>
  <li>A Mesquite file containing your sequences in a matrix. This will need to be <a href="#prepareMatrix">prepared</a>, including setting of codon positions for protein-coding regions.</li>
  </ol>
<p>With these ready basic procedure is to then:</p>
<ol>
  <li><a href="#exporting">Export the matrix and feature table</a> from within Mesquite. This will involve creating or choosing a <a href="#sequenceProfile">sequence profile</a> containing information about the properties of the sequences such as the genetic code, sequence description, etc.</li>
  <li><a href="#usingTbl2asn">Run tbl2asn</a> to create the file to be submitted to GenBank.</li>
</ol>
<h3><a name="downloadTbl2asn"></a>Download tbl2asn and get it ready for use</h3>
<p>To create the file for submission to GenBank, you will need to use the program tbl2asn. Information about this program is available from</p>
<p><a href="https://www.ncbi.nlm.nih.gov/genbank/tbl2asn2/">https://www.ncbi.nlm.nih.gov/genbank/tbl2asn2/</a></p>
<p>Download tbl2asn, and put it in the same directory as the files that you will create in the export process (see below). </p>
<p>You may need to make the file executable. On a UNIX-based machine (MacOSX or Linux), open up a shell (e.g., Terminal on a Mac), cd into the directory containing the file, and then use a command like the following:</p><pre class="monospaced">    chmod +x mac.tbl2asn</pre></p>
<h3><a name="submissionTemplate"></a>Creating the file containing information about the paper</h3>
<p>Go to the <a href="https://submit.ncbi.nlm.nih.gov/genbank/template/submission/">NCBI Submission Portal</a> and create a Submission Template for your paper. This will contain information about your submission to GenBank. Once you have filled in the form, press the Create Template button at the bottom of the page. That will create a text file containing information about your paper, and that file will be downloaded to your computer. Move that file (which will likely be called &quot;template.sbt.txt&quot;), and move it into the same folder as the fasta files you created. You will only need to make this file once; the one file can be used for all of your submissions for that paper.</p>
<h3><a name="sampleTable"></a>Creating the sample table</h3>
<p>The form of the tab-delimited OTU (specimen voucher) ID code file containing the information about the samples is as follows. The first row
  must begin with the word &quot;code&quot;, followed by a tab-delimited list
  of the official GenBank names of the fields that appear in each of the later
  lines. For example,
  if the fields to be included are the organismal name, the taxonomic authority,
  the name of the person who identified the specimen, the country (locality)
  field, the latitude and longitude, and the information identifying the specimen
  voucher, then this first line would appear as follows:</p>
<pre class="monospaced"> code <img src="images/tab.gif" width="8" height="13" /> organism <img src="images/tab.gif" width="8" height="13" /> authority <img src="images/tab.gif" width="8" height="13" /> identified-by <img src="images/tab.gif" width="8" height="13" /> country <img src="images/tab.gif" width="8" height="13" /> lat-lon <img src="images/tab.gif" width="8" height="13" /> specimen-voucher</pre>
<p>(The black triangle, <img src="images/tab.gif" width="8" height="13" />, represents
  a tab.)</p>
<p>The official GenBank names of the fields and the definition of those fields
  is given on the <a href="http://www.ncbi.nlm.nih.gov/Sequin/modifiers.html">Modifiers
    for FASTA Definition Lines</a> page, with more information on the <a href="http://www.ncbi.nlm.nih.gov/Sequin/sequin.hlp.html#ModifiersPage">Sequin
      Help</a> page. You may include whatever fields you need.</p>
<p> On the following lines are the data about  the specimens, one line
  for each specimen. The first item in the line is and ID code. This ID code
  could be your specimen voucher code, or some other unique identifying string.
  You will enter these codes in Mesquite for each sequence, which will allow
  the system to associate the OTU ID code DB information with that particular sequence.
  The following tab-delimited items in the line are the entries for that particular
  specimen. For the example file with the header line shown above, here are two
  lines that contain the information for specimen number 1290, and specimen number
  1633:</p>
<pre class="monospaced">
1290 <img src="images/tab.gif" width="8" height="13" /> Bembidion (Odontium) paraenulum <img src="images/tab.gif" width="8" height="13" /> Bembidion paraenulum Maddison <img src="images/tab.gif" width="8" height="13" /> David Maddison <img src="images/tab.gif" width="8" height="13" /> 
    USA: Mississippi: Walthall Co., Tylertown <img src="images/tab.gif" width="8" height="13" /> 31.0414 N 90.1922 W <img src="images/tab.gif" width="8" height="13" /> Personal: DRMaddison : DNA1290
1633 <img src="images/tab.gif" width="8" height="13" /> Bembidion (Odontium) aenulum <img src="images/tab.gif" width="8" height="13" /> Bembidion aenulum Hayward <img src="images/tab.gif" width="8" height="13" /> David Maddison <img src="images/tab.gif" width="8" height="13" /> 
    USA: Iowa: Jones Co., Oxford Junction <img src="images/tab.gif" width="8" height="13" /> 41.99133 N 91.00671 W <img src="images/tab.gif" width="8" height="13" /> Personal: DRMaddison : DNA1633</pre>
<p>Each of these lines is shown extending over two lines, but that is only for ease of display on this web page. There are thus a total of three lines in this example OTU ID Code DB file:</p>
<pre class="monospaced">
code <img src="images/tab.gif" width="8" height="13" /> organism <img src="images/tab.gif" width="8" height="13" /> authority <img src="images/tab.gif" width="8" height="13" /> identified-by <img src="images/tab.gif" width="8" height="13" /> country <img src="images/tab.gif" width="8" height="13" /> lat-lon <img src="images/tab.gif" width="8" height="13" /> specimen-voucher
1290 <img src="images/tab.gif" width="8" height="13" /> Bembidion (Odontium) paraenulum <img src="images/tab.gif" width="8" height="13" /> Bembidion paraenulum Maddison <img src="images/tab.gif" width="8" height="13" /> David Maddison <img src="images/tab.gif" width="8" height="13" /> 
    USA: Mississippi: Walthall Co., Tylertown <img src="images/tab.gif" width="8" height="13" /> 31.0414 N 90.1922 W <img src="images/tab.gif" width="8" height="13" /> Personal: DRMaddison : DNA1290
1633 <img src="images/tab.gif" width="8" height="13" /> Bembidion (Odontium) aenulum <img src="images/tab.gif" width="8" height="13" /> Bembidion aenulum Hayward <img src="images/tab.gif" width="8" height="13" /> David Maddison <img src="images/tab.gif" width="8" height="13" /> 
    USA: Iowa: Jones Co., Oxford Junction <img src="images/tab.gif" width="8" height="13" /> 41.99133 N 91.00671 W <img src="images/tab.gif" width="8" height="13" /> Personal: DRMaddison : DNA1633</pre>
<h3><a name="prepareMatrix"></a>Preparing the matrix and connecting the taxa to their metadata</h3>
<p>For each matrix to be submitted, you will need to annotated the sequences appropriately. </p>
<p>In particular, the codon positions will need to be specified if any of the sequence are protein-coding. You can specify the codon positions in the List of Characters window in Mesquite. This does not necessarily require that the sequences be aligned, but it will be easier if they are. If they are not aligned, then each will need to be in the correct reading frame (which you can accomplish using the &quot;Alter&gt;Shift to Minimize Stop Codons&quot; tool in Mesquite's matrix editor after you have specified the codon positions in general. </p>
<p>In addition, if there are distinct regions in the fragment that need to be specified in your GenBank submission (e.g., exons and introns, or multiple genes with intervening non-transcribed regions), then you will need to specify these using  Character Groups, which you can do in Mesquite's List of Characters window.</p>
<p>You will also need to specify which taxon in your matrix corresponds to which entry in your sample table file. To do this, go to the Taxa List
  Window (Taxa&amp;Tree&gt;List
  of
  Taxa). You will need to show two new columns in this table. Choose Columns&gt;OTU
  Database and Columns&gt;OTU ID Code. This will show those two columns. Select
  the entire table (with Select All), and touch on the title of the OTU ID code 
  DB column. A menu will appear in which you can choose to browse for your tab-delimited
  OTU ID code DB file. Select that file. The OTU ID code  DB column should indicate
  which
  OTU ID code  DB file to use for each sequence. (In this example, all sequences
  are using the same OTU ID code  DB file.) </p>
<p>Now you need to enter into the OTU code  ID field the ID codes for each of the
  sequences. In the example here, the OTU ID code for the first sequence
  is 1290. This tells Mesquite to look
  in the OTU ID code DB file for the line whose code is 1290 to get the OTU
  information for that sequence. To enter ID codes, use the editing tool (<img src="images/ibeam.png" width="28" height="28" align="absmiddle" />) select the entry, or select the sequence and
  use the popup menu that appears when you touch on the OTU ID Code title at the
  top of the column. Once you have entered all of the OTU ID codes, the Taxa List Window should look something
  like this:</p>
<p align="center"><img src="images/taxonListVoucher.png" width="468" height="226" /></p>
<h3><a name="exporting"></a>Creating the FASTA file and feature table file needed by tbl2asn</h3>
<p>At this point, you are ready to export the files needed by tbl2asn to create your GenBank submission. </p>
<p>Do this by choose File&gt;Export,
  and in the dialog that appears, choose &quot;FASTA (DNA/RNA) for GenBank (tbl2asn)&quot;.
  You will be queried for options:</p>
<p align="center"><img src="images/ExportTbl2asnDialog.png" width="511" height="402" alt=""/></p>
<p>In this dialog box, you will need to choose a sequence profile and various other options. If you haven't created the relevant sequence profile, you should do that first. Sequence profiles are described in the following sections.</p>
<p>In addition to specifiying a sequence profile, you will need to specify various aspects of how Chromaseq formulates  entries. </p>
<p>If you wish to have the OTU ID code added to the DEFINITION line (between the organismal specification and the description of the sequence), then check the &quot;add OTU (Specimen) ID Code to DEFINITION&quot; option. If you have that option checked, then the text you include in the next field will be included immediately in front of the ID code.</p>
<p>A GenBank submission requires that each submission be specified by a &quot;seqID&quot;, a text string that must be unique for each sequence to be submitted. By default, Mesquite uses the taxon name, to which  the gene fragment name can be appended, as the seqID. However, GenBank will automatically truncate the seqID if it is too long. If you have some important text at the end of the seqID that you don't want removed by GenBank, then yoiu can check the &quot;if seqID is too long, shorten in middle&quot; option.</p>
<p>If you are using character gropus to specify the product names for regions of your submission, you should check &quot;use character group name as product name for coding regions&quot; option.</p>
<h3><a name="sequenceProfile"></a>Creating a sequence profile</h3>
<p>A sequence profile stores information about a particular gene, gene fragment, or other region of DNA that you might wish to submit to GenBank. For example, if you have sequenced a region of the CAD gene, you might wish to create a sequence profile for that region. The metadata you include in the sequence profile will then be included along with the rest of your GenBank submission. This sequence profile is then saved by Chromaseq, and the next time you wish to submit sequences of that fragment, you can simply choose that profile from the pull-down menu.</p>
<p>To create a sequence profile, first press the Manage button in the main dialog box. The Sequence Profile Manager will appear:</p>
<p align="center"><img src="images/sequenceProfileManager.png" width="539" height="336" alt=""/></p>
<p>To create a new profile, press &quot;New...&quot;; to edit the contents of a previously created profile, select it in the list and press &quot;Edit...&quot;. The other buttons should be self-explanatory.</p>
<p>By pressing New or Edit, you will be presented with the dialog box in which you can enter data for the sequence profile.</p>
<p align="center"><img src="images/sequenceProfileDialog.png" width="717" height="363" alt=""/></p>
<p>You will need to enter data for the following fields:</p>
<p><em>Description of sequence</em>: the text you enter in this field will be appended to the end of the DESCRIPTION field of GenBank</p>
<p><em>Location</em>: This is the location of the source of the molecular data. The options are:</p>
<ul><li>Genomic</li> 
<li>Mitochondrion</li> 
<li>Chloroplast</li> 
<li>Apicoplast</li> 
<li>Chromatophore</li> 
<li>Chromoplast</li> 
<li>Cyanelle</li> 
<li>Endogenous_virus</li> 
<li>Extrachromosomal</li> 
<li>Hydrogenosome</li> 
<li>Kinetoplast</li> 
<li>Leucoplast</li> 
<li>Macronuclear</li> 
<li>Nucleomorph</li> 
<li>Plasmid</li> 
<li>Plastid</li> 
<li>Proplastid</li> 
<li>Proviral</li> 
</ul>
<p><em>Genetic Code</em>: You will need to specify the genetic code for any protein-coding sequences. For non-coding regions, you can leave this to &quot;The Standard Code&quot;.</p>
<p><em>Molecular type</em>: This is the type of molecular data being submitted. The options are:	</p>
<ul>
<li>genomic DNA </li>
<li>genomic RNA </li>
<li>mRNA </li>
<li>tRNA </li>
<li>rRNA </li>
<li>other RNA </li>
<li>other DNA </li>
<li>transcribed RNA </li>
<li>viral cRNA </li>
<li>unassigned DNA </li>
<li>unassigned RNA</li>
</ul>
<p><em>SeqID suffix</em>: This should be short, and should uniquely distinguish the sequences in one matrix from those in a second matrix that you might submit. For example, it could be the gene name. This will be appended onto each taxon name to form the SeqID for the sequences in the matrix to be submitted.</p>
<p>The optional fields are:</p>
<p><em>Product name</em>: If the entire fragment to be submitted includes genes for only one product, then you can include the product name here. As an alternative, you could specify character groups (see above) and check the &quot;use character group name as product name for coding regions&quot; option in the main export dialog box. For protein-coding regions, you must specify a product name one of these two ways.</p>
<p><em>Note</em>: If you wish to include a note with each of the entries, then enter it in the notes field. For example, you could give details about the fragment of the gene sequenced, or other metadata of relevance to your project.</p>
<p>Once you are ready, press the Export button. We suggest that the file name you ask to create not include any spaces.</p>
<h3>Sharing sequence profiles</h3>
<p>The sequence profiles are stored in the &quot;SequenceProfilesForGenBank&quot; directory that is in the &quot;Mesquite_Prefs&quot; directory of your &quot;Mesquite_Support_Files&quot; directory. You can put copies of any of these files into the equivalent place on another computer to make those profiles available to Mesquite on that other computer.</p>
<h3><a name="usingTbl2asn"></a>Using tbl2asn</h3>
<ul>
  <li>On the Mac or Linux, open up a UNIX shell. (On the Mac, you can start the application called Terminal in the Utilities folder in the Applications folder.) On Windows, open up the command line tool. </li>
  <li>cd into the directory containing tbl2asn and all of the files.</li>
  <li>type one of the following commands and hit Return, where you substitute &quot;fileName.fsa&quot; with the name of the FASTA file:
    <ul>
      <li>MacOSX: <strong>./mac.tbl2asn -t template.sbt.txt -a s2 -V vb -i fileName.fsa</strong></li>
      <li>Windows: <strong>tbl2asn.exe -t template.sbt.txt -a s2 -V vb -i fileName.fsa</strong></li>
      <li>Linux: <strong>./tbl2asn -t template.sbt.txt -a s2 -V vb -i fileName.fsa</strong></li>
    </ul>
  </li>
</ul>
<p>If you wish to see more details about the command-line options available for tbl2asn, see <a href="https://www.ncbi.nlm.nih.gov/genbank/tbl2asn2/">https://www.ncbi.nlm.nih.gov/genbank/tbl2asn2/</a></p>
<ul>
  <li>once you have run tbl2asn, several files will be produced. Examine the errorsummary.val file for a summary of errors and warnings, and the other file that ends with &quot;.val&quot; (e.g., &quot;fileName.val&quot;) for more detailed information about the errors and warnings. Examining the &quot;.val&quot; files will allow you to fix your files and repeat the process until the errors are eliminated. You will need to resolve all errors before you submit the .sqn file to GenBank.</li>
  <li>If you wish to see the GenBank files as they would be formed, examing the .gbf files in a text editor.</li>
  <li>Once all looks good, you can submit the sequences by sending the .sqn files to gb-sub@ncbi.nlm.nih.gov</li>
</ul>
<h2>&nbsp;</h2>
<!-- InstanceEndEditable -->

    
    
    </div>
	<!-- This clearing element should immediately follow the #mainContent div in order to force the #container div to contain all child floats --><br class="clearfloat" />
  <div id="footer">
    <p> ©2018 David R. Maddison &amp; Wayne P. Maddison</p>
  <!-- end #footer --></div>
<!-- end #container --></div>

<script type="text/javascript">
<!--


var GettingStartedPanel = null;
var ChromatogramViewerPanel = null;
var ProcessingChromatoPanel = null;
var PreparingDataPanel = null;
var CreditPanel = null;
var HelpPanel = null;
var GenBankPanel = null;
var OtherFeaturesPanel = null;
var TechnicalPanel = null;

if (qsParm['ProcessingChromatoPanel']=='open')
 	ProcessingChromatoPanel = new Spry.Widget.CollapsiblePanel("ProcessingChromatoPanel", {contentIsOpen: true});
else
 	ProcessingChromatoPanel = new Spry.Widget.CollapsiblePanel("ProcessingChromatoPanel", {contentIsOpen: false});
	
if (qsParm['GettingStartedPanel']=='open')
 	GettingStartedPanel = new Spry.Widget.CollapsiblePanel("GettingStartedPanel", {contentIsOpen: true});
else
 	GettingStartedPanel = new Spry.Widget.CollapsiblePanel("GettingStartedPanel", {contentIsOpen: false});
	
if (qsParm['ChromatogramViewerPanel']=='open')
 	ChromatogramViewerPanel = new Spry.Widget.CollapsiblePanel("ChromatogramViewerPanel", {contentIsOpen: true});
else
 	ChromatogramViewerPanel = new Spry.Widget.CollapsiblePanel("ChromatogramViewerPanel", {contentIsOpen: false});

if (qsParm['PreparingDataPanel']=='open')
 	PreparingDataPanel = new Spry.Widget.CollapsiblePanel("PreparingDataPanel", {contentIsOpen: true});
else
 	PreparingDataPanel = new Spry.Widget.CollapsiblePanel("PreparingDataPanel", {contentIsOpen: false});

if (qsParm['CreditPanel']=='open')
 	CreditPanel = new Spry.Widget.CollapsiblePanel("CreditPanel", {contentIsOpen: true});
else
 	CreditPanel = new Spry.Widget.CollapsiblePanel("CreditPanel", {contentIsOpen: false});

if (qsParm['HelpPanel']=='open')
 	HelpPanel = new Spry.Widget.CollapsiblePanel("HelpPanel", {contentIsOpen: true});
else
 	HelpPanel = new Spry.Widget.CollapsiblePanel("HelpPanel", {contentIsOpen: false});

if (qsParm['GenBankPanel']=='open')
 	GenBankPanel = new Spry.Widget.CollapsiblePanel("GenBankPanel", {contentIsOpen: true});
else
 	GenBankPanel = new Spry.Widget.CollapsiblePanel("GenBankPanel", {contentIsOpen: false});
    
if (qsParm['OtherFeaturesPanel']=='open')
 	OtherFeaturesPanel = new Spry.Widget.CollapsiblePanel("OtherFeaturesPanel", {contentIsOpen: true});
else
 	OtherFeaturesPanel = new Spry.Widget.CollapsiblePanel("OtherFeaturesPanel", {contentIsOpen: false});
    
if (qsParm['TechnicalPanel']=='open')
 	TechnicalPanel = new Spry.Widget.CollapsiblePanel("TechnicalPanel", {contentIsOpen: true});
else
 	TechnicalPanel = new Spry.Widget.CollapsiblePanel("TechnicalPanel", {contentIsOpen: false});



//-->
</script>
</body>
<!-- InstanceEnd --></html>
