<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><!-- InstanceBegin template="/Templates/ChromaseqTemplate.dwt" codeOutsideHTMLIsLocked="false" -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />


<!-- InstanceBeginEditable name="doctitle" -->
<title>Preparation &amp; Workflow</title>
<!-- InstanceEndEditable -->

<![endif]--><!--[if IE]>
<style type="text/css"> 
/* place css fixes for all versions of IE in this conditional comment */
.thrColFixHdr #sidebar2, .thrColFixHdr #sidebar1 { padding-top: 30px; }
.thrColFixHdr #mainContent { zoom: 1; }
/* the above proprietary zoom property gives IE the hasLayout it needs to avoid several bugs */
</style>
<![endif]-->

<style type="text/css">
<!--
-->
</style>
<link href="CSS/mesquitePackage.css" rel="stylesheet" type="text/css" />
<link href="SpryAssets/SpryCollapsiblePanel.css" rel="stylesheet" type="text/css" />

<script src="SpryAssets/SpryCollapsiblePanel.js" type="text/javascript"></script>
<script type="text/javascript" src="SpryAssets/SpryURLUtils.js"></script>
<script type="text/javascript" src="javascript/mesquiteChromaseq.js"></script>

<!-- InstanceBeginEditable name="head" -->
<style type="text/css">
-->
</style><!-- InstanceEndEditable -->
</head>

<body bgcolor="#f4f0e2" class="thrColFixHdr" >

<div id="container">
  <div id="header">
    <table width="100%" border="0" cellpadding="4">
  <tr>
    <td width="51%"><strong id="PackageTitle">Chromaseq</strong></td>
    <td width="33%" align="right"><div id="subtitle"><!-- InstanceBeginEditable name="subtitle" -->Preparation &amp; Processing<!-- InstanceEndEditable --></div></td>
    <td width="2%">&nbsp;</td>
    <td width="14%"><p>Version 1.5 <br />
      27 December
2018<br />
      </p></td>
  </tr>
</table>

</div>
  
  
<div id="sidebar1">
  
  
  
  
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('index.html')">Home</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('overview.html')">Overview</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('download.html')">Download</a>
  </div>

  <div id="GettingStartedPanel" class="CollapsiblePanel">
    <div class="CollapsiblePanelTab" tabindex="0">Getting Started ˇ</div>
    <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('download.html');">Download</a></td></tr>
  			<tr><td><a href="javascript:pageLink('installation.html');">Installation</a></td></tr>
  			<tr><td><a href="javascript:pageLink('quickstart.html');">Quick Start</a></td></tr>
  			<tr><td><a href="javascript:pageLink('tutorial.html');">Tutorial</a></td></tr>
		</table>
    </div>
  </div>
  <div id="PreparingDataPanel" class="CollapsiblePanel">
    <div class="CollapsiblePanelTab" tabindex="0">Preparing Data ˇ</div>
    <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('chromatograms.html');">Chromatograms</a></td></tr>
  			<tr><td><a href="javascript:pageLink('fileNamingRules.html');">File Naming Rules</a></td></tr>
  			<tr><td><a href="javascript:pageLink('sampleCode.html');">Samples & Sequence Names</a></td></tr>
  			<tr><td><a href="javascript:pageLink('primers.html');">Primer & Gene Information</a></td></tr>
		</table>
    </div>
  </div>
  
<div id="ProcessingChromatoPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Processing Chromatograms ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr> <td><a href="javascript:pageLink('preparation.html');">Preparation & Processing</a></td></tr>
  			<tr><td><a href="javascript:pageLink('phredPhrap.html');">Phred & Phrap Locations & Options</a></td></tr>
  			<tr><td><a href="javascript:pageLink('postPhrap.html');">Post-Phrap Options</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('quality.html');">Quality Scores & Colors</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('segregatingChromatograms.html');">Segregating Chromatogram Files</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('appendingSequences.html');">Processing Additional Chromatograms</a></td></tr>
		</table>
  </div>
</div>


<div id="ChromatogramViewerPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Chromatogram Viewer ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('viewer.html');">Viewing Chromatograms</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('callingBases.html');">Altering Base Calls</a></td></tr>
  			<tr><td><a href="javascript:pageLink('conflicts.html');">Detecting Conflicts</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('primerSites.html');">Displaying Primer Sites</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('extraSequences.html');">Multiple Reads/Contigs at Once</a></td></tr>
  			<tr><td><a href="javascript:pageLink('display.html');">Other Display Options</a></td></tr>
			<tr><td><a href="javascript:pageLink('navigation.html');">Navigation</a></td></tr>
		</table>
  </div>
</div>

<div id="GenBankPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">GenBank Submission ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('genbank.html');">GenBank via tbl2asn</a></td></tr>
  			<tr><td><a href="javascript:pageLink('genbankSequin.html');">GenBank via Sequin</a></td></tr>
		</table>
  </div>
</div>

<div id="OtherFeaturesPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Other Features ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('listWindow.html');">List Window Features</a></td></tr>
  			<tr><td><a href="javascript:pageLink('matrixCells.html');">Matrix Colors</a></td></tr>
  			<tr><td><a href="javascript:pageLink('purging.html');">Purging Chromaseq Data</a></td></tr>
  			<tr><td><a href="javascript:pageLink('helpfulMesquite.html');">Helpful Mesquite Features</a></td></tr>
		</table>
  </div>
</div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('workflow.html')">Workflow Suggestions</a>
  </div>

<div id="HelpPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Help & Info ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('tutorial.html');">Tutorial</a></td></tr>
  			<tr><td><a href="javascript:pageLink('faq.html');">FAQ</a></td></tr>
  			<tr><td><a href="javascript:pageLink('troubleshooting.html');">Troubleshooting</a></td></tr>
  			<tr><td><a href="javascript:pageLink('chromaseqList.html');">Chromaseq E-mail List</a></td></tr>
  			<tr><td><a href="javascript:pageLink('limitations.html');">Limitations</a></td></tr>
  			<tr><td><a href="javascript:pageLink('otherPrograms.html');">Chromaseq &amp; Other Programs</a></td></tr>
  			<tr><td><a href="javascript:pageLink('future.html');">Future Directions</a></td></tr>
	</table>
  </div>
</div>

<div id="TechnicalPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Technical Details ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('sourceCode.html');">Source Code</a></td></tr>
  			<tr><td><a href="javascript:pageLink('technicalFAQ.html');">Technical FAQ</a></td></tr>
		</table>
  </div>
</div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('knownIssues.html')">Known Issues</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('versions.html')">Versions</a>
  </div>

<div id="CreditPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Citation & Credit ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('citation.html');">How to Cite</a></td></tr>
  			<tr><td><a href="javascript:pageLink('acknowledgements.html');">Acknowledgements</a></td></tr>
		</table>
  </div>
</div>


<h3>&nbsp;</h3>
  <!-- end #sidebar1 --></div>

  <div id="mainContent">
    
<!-- InstanceBeginEditable name="content" -->
<h2>Preparing for Chromatogram Processing</h2>
<p>Before beginning to process chromatograms, make sure you have:</p>
<ol>
  <li><a href="javascript:pageLink('installation.html');">Installed Chromaseq, Phred, and Phrap</a></li>
  <li><a href="javascript:pageLink('chromatograms.html');">Prepared the chromatograms</a></li>
  <li>Prepared the <a href="javascript:pageLink('sampleCodes.html');">sample codes file</a> and the <a href="javascript:pageLink('primers.html');">primer information file</a></li>
</ol>
<h2>The first step in processing chromatograms</h2>
<p>There are several ways Chromaseq can process chromatograms. You should first determine which of the following meet your needs, and begin the process by choosing the menu item mentioned:</p>
<ul>
  <li>Process all of the chromatogram files in a directory, making base calls, contigs, and then creating matrices in a Mesquite project, one matrix for each gene, and filling the matrices with the sequences generated. This option is available under <strong>File&gt;Open Other&gt;Process Chromatogram Files in Directory</strong>.</li>
  <li>Process all of the chromatogram files in a directory, making base calls, contigs, and then appending the sequences to existing matrices in a Mesqutie project. If a matrix for a particular gene fragment doesn't exist in the project, it will be created. This option is available under <strong>Characters&gt;Append Sequences from Chromatograms</strong></li>
  <li>Process all of the chromatogram files in a directory, making base calls, contigs, but not reading the results in. This option is available under <strong>Utilities&gt;Process Chromatograms</strong>.  The Utilities menu is available when the Mesquite log window is frontmost.</li>
  <li>Segregate (into a subdirectory) all of the chromatogram file in a directory that meet specified critieria (e.g., that belong to a certain taxon, or that are part of a specified gene). This doesn't make base calls or contigs, it simply moves the specified files into a subdirectory. This option is available under <strong>Utilities&gt;Segregate Chromatograms</strong>.  The Utilities menu is available when the Mesquite log window is frontmost.</li>
</ul>
<h2>Your next choices</h2>
<p>Once you ask to start the processing, you will be asked to make some choices before you get to the main processing dialog box. In particular, you will be asked to </p>
<p>1. Choose a <a href="javascript:pageLink('fileNamingRules.html');">file naming rule</a>; if the appropriate naming rule hasn't already been created, you will need to <a href="javascript:pageLink('fileNamingRules.html#defineNamingRules');">create one</a>.</p>
<p align="center"><img src="images/choosingNamingRule.png" width="349" height="138" /></p>
<p>2. Select a method to determine <a href="javascript:pageLink('sampleCodes.html');">sequence names</a>. You will be asked to choose among the two options available:</p>
<p align="center"><img src="images/sequenceNameSupplier.jpg" width="299" height="356" /></p>
<p>The only choices currently offered are to look up the names for sequences in a <a href="javascript:pageLink('sampleCode.html');">sample codes and sequence names file</a>, or to simply use the sample code contained in the file name as the sequence name. If you choose to have the sequence names looked up in a file, you will be asked for the location of the file:</p>
<p align="center"><img src="images/sampleCodeFIleLocation.png" width="384" height="127" /></p>
<p>&nbsp;</p>
<p>3. Select a text file that contains <a href="javascript:pageLink('primers.html');">information about primers and genes</a>:</p>
<p align="center"><img src="images/primerInfoFileLocation.png" width="393" height="127" /></p>
<p>These dialog boxes will only appear the first time you use Chromaseq. In later uses, Chromaseq will remember your choices, and will automatically use them as defaults. However, you do have an opportunity to change your choices, should you wish, in the Phred/Phrap/Chromaseq Options dialog box, described in the next section. </p>
<h2><a name="processing" id="processing"></a>The Processing Options dialog box</h2>
<p>At this point, you will be presented with the main dialog for processing chromatograms. After setting up the options, and pressing &quot;Process&quot;, Chromaseq will start processing the chromatograms. Details of that process are described next, but first you need to set the options appropriately.</p>
<p align="center"><img src="images/phredPhrapDialogQuickstart.png" width="391" height="501" /></p>
<p>The top of this dialog also lists the choices you have just made about how file names should be processed, the source of names for sequences, and the source of primer information. There are also buttons beside these elements that allow you to alter your choices, in the event that you change your mind.</p>
<p>If you haven't already, you will need to tell Chromaseq where the helper programs are. You can do this by touching on the &quot;Phred Phrap Locations &amp; Options&quot; button; details are on the <a href="javascript:pageLink('phredPhrap.html');">Phred &amp; Phrap Locations &amp; Options</a> page.  On the same page are options within Phred and Phrap that you can set.</p>
<p>In addition, you have the following choice of options that involve preparation before the Phred and Phrap processing, and that involve the process itself:</p>
<ul>
  <li>Choose whether to only process files with standard extensions (such as &quot;.ab1&quot; or &quot;.CRO&quot;). If you turn this option off, then Chromaseq will process all files in the directory, which will be problematic if you have additional files (such as &quot;.seq&quot; files) in the same directory.</li>
  <li>Choose whether to have the extension of the chromatograms changed to a specified extension when they are organized into their subdirectories. If you enter text in the &quot;uniform file extension for organized chromatograms (if desired)&quot; box, then any old extension to the file names will be removed, and they will be replaced by the extension you specify. If you don't specify any extension, the extension on the original file names will be used. We frankly can't remember why this feature was added, but we have left it in as we suspect there are some contexts in which this is valuable.</li>
  <li>Choose whether to have Chromaseq save a separate directory with copies of all of your original chromatograms in there. By default, Chromaseq will move all of the chromatograms into subdirectories, and rename the files in the process. If you wish to have backups saved of the chromatograms with their original file names, then check this box or save backups in some other way.</li>
  <li>Choose the <a href="javascript:pageLink('phredPhrap.html');">location and options for Phred and Phrap</a>, as well as the location for the phredpar.dat file.</li>
  </ul>
<p>You may also choose to alter the options for how Chromaseq processes the sequences after Phred and Phrap do their thing:</p>
<ul>
  <li><a href="javascript:pageLink('postPhrapOptions.html');">Post-Phrap Sequence Processing Options</a>: This dialog box specifies how Chromaseq processes the contigs and sequences after Phred and Phrap are done with them. By default, Chromaseq trims the sequences, searches for multiple peaks that should be given ambiguity codes (e.g., R or Y), etc.</li>
  <li>Choose whether or not a single or multiple taxon blocks should be created if there are multiple genes represented in the sequences. By default, Chromaseq creates a separate taxon block for each gene. However, if you check this option, only one taxon block will be created that will includes as taxa all of the distinct sequence names. This option is only available if you are importing the results into a a Mesquite project, i.e., if you have chosen <strong>File&gt;Open Other&gt;Process Chromatogram Files in Directory</strong>.</li>
  <li>Choose whether matrices that are created in Mesquite are opened up in a bird's eye view, or instead in the normal view.</li>
</ul>
<p>&nbsp;</p>
<h2><a name="duringProcessing" id="duringProcessing"></a>What happens during chromatogram processing</h2>
<p>Once you press the Process button, the following things happen:</p>
<ol>
  <li>If you choose to save backups of the chromatograms, a directory called &quot;originals&quot; will be created, and copies of all of the chromatograms will be placed in there as backups.</li>
  <li>Chromaseq creates a &quot;processed&quot; directory in the directory containing the chromatograms. </li>
  <li>Chromaseq scans the chromatogram file names, finds the primer names contained within each, looks up the gene fragment each corresponds to, and accumulates a list of all available gene fragments represented among the chromatograms. It then creates one subdirectory within the &quot;processed&quot; directory for each gene fragment. </li>
  <li>For each gene fragment, it scans the corresponding chromatogram files and find all the files with the same one sample code. It creates a subdirectory in that gene fragment directory for that sample, names it according to the short sequence name, and moves all of the chromatogram files for that sample and that gene fragment into the subdirectory. In so doing, it renames the chromatogram files so that they meet Phred's requirements.</li>
  <li>Once the chromatograms are renamed and moved into their subdirectories, Chromaseq will create a shell script that will start up Phred and Phrap and ask them to process all of the chromatograms in each subdirectory. It will then invoke that shell script in the shell program of relevance for your computer's operating system. On a Mac, this
    will typically be the Terminal application; on Windows, it will be the Cygwin
    shell if that is what you installed; on a Linux variant, it will be the bash shell or something similar.</li>
  <li>Once Phred and Phrap are invoked, you should see a lot of text whizzing by in the shell program. If at this point you go back into Mesquite, and look in the log window, you will see that Mesquite is reporting on the progress of the analysis. For example, it might look something like this:</li>
</ol>
<p class="monospacedSmall">Phred Phrap processing of chromatograms as scripted by Mesquite<br />
  Processing directory: /ChromaseqExamples/Chromatograms</p>
<p class="monospacedSmall">Using names and codes file: /ChromaseqExamples/SampleCodes.txt</p>
<p class="monospacedSmall">Using primers file: /ChromaseqExamples/Primers.txt</p>
<p class="monospacedSmall"><span class="monospacedSmall">Parameters: Phr</span>ed Options:  -id . -pd . -d <br />
  Phrap Options:  -new_ace -qual_show 20 -vector_bound 0 <br />
  Chromaseq Options: <br />
  Quality threshold for lower case: 49<br />
  Convert multiple-peaks sites to ambiguity codes: true<br />
  Minimum secondary peak fraction for ambiguity: 0.3<br />
  Trim low quality regions from ends: true<br />
  Quality threshold for trim: 20<br />
  Trim window length: 10<br />
  Trim window threshold: 5</p>
<p class="monospacedSmall">---------------------------------------------------<br />
  Section 1: Examining chromatogram files and preparing directories </p>
<p class="monospacedSmall"> Preparing Bemb perspicuum AZ 1120  (COI)<br />
  Preparing Bemb louisella NS 1250  (COI)<br />
  Preparing Bemb louisella NS 1250  (28S)<br />
  Preparing Bemb perspicuum CA 2320  (28S)<br />
  Preparing Bemb paraenulum VA 1856  (COI)<br />
  Preparing Bemb perspicuum AZ 1120  (28S)<br />
  Preparing Bemb paraenulum VA 1856  (28S)</p>
<p class="monospacedSmall">Number of files prepared: 7</p>
<p class="monospacedSmall">---------------------------------------------------<br />
  Section 2: Renaming chromatograms for Phred and Phrap</p>
<p class="monospacedSmall"> A01_A01DNA1120_B1490_411940.ab1  renamed to  DNA1120.b.B1490.ab1<br />
  A01_A01DNA1120_Bcoi2R_412036.ab1  renamed to  DNA1120.g.Bcoi2R.ab1<br />
  C07_C07DNA1250_B1490_208508.ab1  renamed to  DNA1250.b.B1490.ab1<br />
  C07_C07DNA1250_B2232_208604.ab1  renamed to  DNA1250.g.B2232.ab1<br />
  C12_C12DNA1250_D3aR_572645.ab1  renamed to  DNA1250.g.D3aR.ab1<br />
  C12_C12DNA1250_NLF184_572549.ab1  renamed to  DNA1250.b.NLF184.ab1<br />
  F06_F06DNA2320_D3aR_552091.ab1  renamed to  DNA2320.g.D3aR.ab1<br />
  F06_F06DNA2320_NLF184_551995.ab1  renamed to  DNA2320.b.NLF184.ab1<br />
  H05_H05DNA1856_B1490_237723.ab1  renamed to  DNA1856.b.B1490.ab1<br />
  H05_H05DNA1856_B2232_237819.ab1  renamed to  DNA1856.g.B2232.ab1<br />
  H07_H07DNA1120_D1_394061.ab1  renamed to  DNA1120.b.D1.ab1<br />
  H07_H07DNA1120_D3_394157.ab1  renamed to  DNA1120.g.D3.ab1<br />
  H09_H09DNA1856_D1_237563.ab1  renamed to  DNA1856.b.D1.ab1<br />
  H09_H09DNA1856_D3_237659.ab1  renamed to  DNA1856.g.D3.ab1</p>
<p class="monospacedSmall">---------------------------------------------------<br />
  Section 3: Execution of Phred-Phrap command file begun</p>
<p class="monospacedSmall">.1. Bemb perspicuum AZ 1120 (Phred)<br />
  2. Bemb louisella NS 1250 (Phrap)<br />
  3. Bemb louisella NS 1250 (Phrap)<br />
  4. Bemb perspicuum CA 2320 (Phrap)<br />
  6. Bemb perspicuum AZ 1120 (Phred)</p>
<p class="monospacedSmall">Phred &amp; Phrap analyses completed.</p>
<p class="monospacedSmall">---------------------------------------------------<br />
  Section 4: Mesquite processing of Phrap results<br />
  <br />
  Processing ACE file: Bemb louisella NS 1250.ace<br />
  C12_C12DNA1250_D3aR_572645.ab1 (rev)<br />
  C12_C12DNA1250_NLF184_572549.ab1</p>
<p class="monospacedSmall">Processing ACE file: Bemb paraenulum VA 1856.ace<br />
  H09_H09DNA1856_D3_237659.ab1 (rev)<br />
  H09_H09DNA1856_D1_237563.ab1</p>
<p class="monospacedSmall">Processing ACE file: Bemb perspicuum AZ 1120.ace<br />
  H07_H07DNA1120_D3_394157.ab1 (rev)<br />
  H07_H07DNA1120_D1_394061.ab1</p>
<p class="monospacedSmall">Processing ACE file: Bemb perspicuum CA 2320.ace<br />
  F06_F06DNA2320_D3aR_552091.ab1 (rev)<br />
  F06_F06DNA2320_NLF184_551995.ab1</p>
<p class="monospacedSmall">Processing ACE file: Bemb louisella NS 1250.ace<br />
  C07_C07DNA1250_B2232_208604.ab1 (rev)<br />
  C07_C07DNA1250_B1490_208508.ab1</p>
<p class="monospacedSmall">Processing ACE file: Bemb paraenulum VA 1856.ace<br />
  H05_H05DNA1856_B2232_237819.ab1 (rev)<br />
  H05_H05DNA1856_B1490_237723.ab1</p>
<p class="monospacedSmall">Processing ACE file: Bemb perspicuum AZ 1120.ace<br />
  A01_A01DNA1120_Bcoi2R_412036.ab1 (rev)<br />
  A01_A01DNA1120_B1490_411940.ab1</p>
<p class="monospacedSmall">========================<br />
  Section 5: Phred/Phrap processing report</p>
<p class="monospacedSmall">Multi-read sequences obtained from the following genes and samples:<br />
  (28S) -- Bemb louisella NS 1250<br />
  (28S) -- Bemb paraenulum VA 1856<br />
  (28S) -- Bemb perspicuum AZ 1120<br />
  (28S) -- Bemb perspicuum CA 2320<br />
  (COI) -- Bemb louisella NS 1250<br />
  (COI) -- Bemb paraenulum VA 1856<br />
  (COI) -- Bemb perspicuum AZ 1120</p>
<p class="monospacedSmall">========================</p>
<p class="monospacedSmall">Processing of chromatogram files completed.</p>
<pre class="monospacedSmall">&nbsp;</pre>
<p>If it doesn't look something like this, or if the shell program has very little text that appeared in it, then something might have gone wrong. Check to make sure you followed the previous steps carefully. If it still doesn't seem to be working, you should check out the <a href="troubleshooting.html">Troubleshooting</a> page. </p>
<ol start="7">
  <li>Once Phred and Phrap have finished processing, Chromaseq will then examine each of the contigs produced by Phrap, and alter them as it deems appropriate, according to the options that were specified in the Post-Phrap Sequence Processing Options dialog box. For example, Phred and Phrap will never make a base call that is ambiguous. If there are two clear peaks at a particular site, representing both A and G, Phred and Phrap will always choose to call the base either A or G, never R. Chromaseq by default will look through the chromatograms for evidence of multiple peaks at a site, and change Phrap's base calls to ambiguity codes if there are. Chromaseq will also trim low-quality ends from the sequences. Chromaseq will keep a copy of Phred and Phrap's original contigs with the original calls (in the &quot;.ace&quot; file contained within the sequences subdirectory) and make a new &quot;m.ace&quot; file with the modified contigs. More details are on the <a href="javascript:pageLink('postPhrapOptions.html');">Post-Phrap Sequence Processing Options</a> page. </li>
  <li>If you have chosen to import the results into a Mesquite project (i.e., if you chose  <strong>File&gt;Open Other&gt;Process Chromatogram Files in Directory</strong>), then Chromaseq will create a new Mesquite project, containing one matrix per gene fragment, and it will read in the sequences into those matrices; the matrices will be displayed with the <a href="javascript:pageLink('matrixCells.html');">color of the cells</a> being related to the quality score of the base as judged by Phred and Phrap.</li>
  <li>Chromaseq will also produce a directory containing FASTA files, one for each successfully contig'd sequence, that will allow you to examine the results quickly. You can, for example, drop these files into another Mesquite matrix, which will allow you to add those sequences to another matrix.</li>
</ol>
<p>At this point, if you wish, you can examine the chromatograms for a sequence using the <a href="javascript:pageLink('viewer.html');">Chromatogram Viewer</a>.</p>
<p>You can also align the sequences, reverse complement them, etc. As long as you do this from within Mesquite, you should still be able to examine the chromatograms for a particular sequence and make base calls. However, if you make these alternations by saving the sequences to a file, opening up that file in another alignment program, then importing the alignment back into Mesquite, for example, the connection between the sequence and the chromatograms will be lost, and there is currently no way to reconnect them.</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<!-- InstanceEndEditable -->

    
    
    </div>
	<!-- This clearing element should immediately follow the #mainContent div in order to force the #container div to contain all child floats --><br class="clearfloat" />
  <div id="footer">
    <p> ©2018 David R. Maddison &amp; Wayne P. Maddison</p>
  <!-- end #footer --></div>
<!-- end #container --></div>

<script type="text/javascript">
<!--


var GettingStartedPanel = null;
var ChromatogramViewerPanel = null;
var ProcessingChromatoPanel = null;
var PreparingDataPanel = null;
var CreditPanel = null;
var HelpPanel = null;
var GenBankPanel = null;
var OtherFeaturesPanel = null;
var TechnicalPanel = null;

if (qsParm['ProcessingChromatoPanel']=='open')
 	ProcessingChromatoPanel = new Spry.Widget.CollapsiblePanel("ProcessingChromatoPanel", {contentIsOpen: true});
else
 	ProcessingChromatoPanel = new Spry.Widget.CollapsiblePanel("ProcessingChromatoPanel", {contentIsOpen: false});
	
if (qsParm['GettingStartedPanel']=='open')
 	GettingStartedPanel = new Spry.Widget.CollapsiblePanel("GettingStartedPanel", {contentIsOpen: true});
else
 	GettingStartedPanel = new Spry.Widget.CollapsiblePanel("GettingStartedPanel", {contentIsOpen: false});
	
if (qsParm['ChromatogramViewerPanel']=='open')
 	ChromatogramViewerPanel = new Spry.Widget.CollapsiblePanel("ChromatogramViewerPanel", {contentIsOpen: true});
else
 	ChromatogramViewerPanel = new Spry.Widget.CollapsiblePanel("ChromatogramViewerPanel", {contentIsOpen: false});

if (qsParm['PreparingDataPanel']=='open')
 	PreparingDataPanel = new Spry.Widget.CollapsiblePanel("PreparingDataPanel", {contentIsOpen: true});
else
 	PreparingDataPanel = new Spry.Widget.CollapsiblePanel("PreparingDataPanel", {contentIsOpen: false});

if (qsParm['CreditPanel']=='open')
 	CreditPanel = new Spry.Widget.CollapsiblePanel("CreditPanel", {contentIsOpen: true});
else
 	CreditPanel = new Spry.Widget.CollapsiblePanel("CreditPanel", {contentIsOpen: false});

if (qsParm['HelpPanel']=='open')
 	HelpPanel = new Spry.Widget.CollapsiblePanel("HelpPanel", {contentIsOpen: true});
else
 	HelpPanel = new Spry.Widget.CollapsiblePanel("HelpPanel", {contentIsOpen: false});

if (qsParm['GenBankPanel']=='open')
 	GenBankPanel = new Spry.Widget.CollapsiblePanel("GenBankPanel", {contentIsOpen: true});
else
 	GenBankPanel = new Spry.Widget.CollapsiblePanel("GenBankPanel", {contentIsOpen: false});
    
if (qsParm['OtherFeaturesPanel']=='open')
 	OtherFeaturesPanel = new Spry.Widget.CollapsiblePanel("OtherFeaturesPanel", {contentIsOpen: true});
else
 	OtherFeaturesPanel = new Spry.Widget.CollapsiblePanel("OtherFeaturesPanel", {contentIsOpen: false});
    
if (qsParm['TechnicalPanel']=='open')
 	TechnicalPanel = new Spry.Widget.CollapsiblePanel("TechnicalPanel", {contentIsOpen: true});
else
 	TechnicalPanel = new Spry.Widget.CollapsiblePanel("TechnicalPanel", {contentIsOpen: false});



//-->
</script>
</body>
<!-- InstanceEnd --></html>
