<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><!-- InstanceBegin template="/Templates/ChromaseqTemplate.dwt" codeOutsideHTMLIsLocked="false" -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />


<!-- InstanceBeginEditable name="doctitle" -->
<title>Tutorial</title>
<!-- InstanceEndEditable -->

<![endif]--><!--[if IE]>
<style type="text/css"> 
/* place css fixes for all versions of IE in this conditional comment */
.thrColFixHdr #sidebar2, .thrColFixHdr #sidebar1 { padding-top: 30px; }
.thrColFixHdr #mainContent { zoom: 1; }
/* the above proprietary zoom property gives IE the hasLayout it needs to avoid several bugs */
</style>
<![endif]-->

<style type="text/css">
<!--
-->
</style>
<link href="CSS/mesquitePackage.css" rel="stylesheet" type="text/css" />
<link href="SpryAssets/SpryCollapsiblePanel.css" rel="stylesheet" type="text/css" />

<script src="SpryAssets/SpryCollapsiblePanel.js" type="text/javascript"></script>
<script type="text/javascript" src="SpryAssets/SpryURLUtils.js"></script>
<script type="text/javascript" src="javascript/mesquiteChromaseq.js"></script>

<!-- InstanceBeginEditable name="head" -->
<style type="text/css">
<!--
-->
</style>
<!-- InstanceEndEditable -->
</head>

<body bgcolor="#f4f0e2" class="thrColFixHdr" >

<div id="container">
  <div id="header">
    <table width="100%" border="0" cellpadding="4">
  <tr>
    <td width="51%"><strong id="PackageTitle">Chromaseq</strong></td>
    <td width="33%" align="right"><div id="subtitle"><!-- InstanceBeginEditable name="subtitle" -->
    <p>Tutorial</p>
    <!-- InstanceEndEditable --></div></td>
    <td width="2%">&nbsp;</td>
    <td width="14%"><p>Version 1.5 <br />
      27 December
2018<br />
      </p></td>
  </tr>
</table>

</div>
  
  
<div id="sidebar1">
  
  
  
  
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('index.html')">Home</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('overview.html')">Overview</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('download.html')">Download</a>
  </div>

  <div id="GettingStartedPanel" class="CollapsiblePanel">
    <div class="CollapsiblePanelTab" tabindex="0">Getting Started ˇ</div>
    <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('download.html');">Download</a></td></tr>
  			<tr><td><a href="javascript:pageLink('installation.html');">Installation</a></td></tr>
  			<tr><td><a href="javascript:pageLink('quickstart.html');">Quick Start</a></td></tr>
  			<tr><td><a href="javascript:pageLink('tutorial.html');">Tutorial</a></td></tr>
		</table>
    </div>
  </div>
  <div id="PreparingDataPanel" class="CollapsiblePanel">
    <div class="CollapsiblePanelTab" tabindex="0">Preparing Data ˇ</div>
    <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('chromatograms.html');">Chromatograms</a></td></tr>
  			<tr><td><a href="javascript:pageLink('fileNamingRules.html');">File Naming Rules</a></td></tr>
  			<tr><td><a href="javascript:pageLink('sampleCode.html');">Samples & Sequence Names</a></td></tr>
  			<tr><td><a href="javascript:pageLink('primers.html');">Primer & Gene Information</a></td></tr>
		</table>
    </div>
  </div>
  
<div id="ProcessingChromatoPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Processing Chromatograms ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr> <td><a href="javascript:pageLink('preparation.html');">Preparation & Processing</a></td></tr>
  			<tr><td><a href="javascript:pageLink('phredPhrap.html');">Phred & Phrap Locations & Options</a></td></tr>
  			<tr><td><a href="javascript:pageLink('postPhrap.html');">Post-Phrap Options</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('quality.html');">Quality Scores & Colors</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('segregatingChromatograms.html');">Segregating Chromatogram Files</a></td></tr>
  			<tr> <td><a href="javascript:pageLink('appendingSequences.html');">Processing Additional Chromatograms</a></td></tr>
		</table>
  </div>
</div>


<div id="ChromatogramViewerPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Chromatogram Viewer ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('viewer.html');">Viewing Chromatograms</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('callingBases.html');">Altering Base Calls</a></td></tr>
  			<tr><td><a href="javascript:pageLink('conflicts.html');">Detecting Conflicts</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('primerSites.html');">Displaying Primer Sites</a></td></tr>
  			<tr>  <td><a href="javascript:pageLink('extraSequences.html');">Multiple Reads/Contigs at Once</a></td></tr>
  			<tr><td><a href="javascript:pageLink('display.html');">Other Display Options</a></td></tr>
			<tr><td><a href="javascript:pageLink('navigation.html');">Navigation</a></td></tr>
		</table>
  </div>
</div>

<div id="GenBankPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">GenBank Submission ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('genbank.html');">GenBank via tbl2asn</a></td></tr>
  			<tr><td><a href="javascript:pageLink('genbankSequin.html');">GenBank via Sequin</a></td></tr>
		</table>
  </div>
</div>

<div id="OtherFeaturesPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Other Features ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('listWindow.html');">List Window Features</a></td></tr>
  			<tr><td><a href="javascript:pageLink('matrixCells.html');">Matrix Colors</a></td></tr>
  			<tr><td><a href="javascript:pageLink('purging.html');">Purging Chromaseq Data</a></td></tr>
  			<tr><td><a href="javascript:pageLink('helpfulMesquite.html');">Helpful Mesquite Features</a></td></tr>
		</table>
  </div>
</div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('workflow.html')">Workflow Suggestions</a>
  </div>

<div id="HelpPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Help & Info ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('tutorial.html');">Tutorial</a></td></tr>
  			<tr><td><a href="javascript:pageLink('faq.html');">FAQ</a></td></tr>
  			<tr><td><a href="javascript:pageLink('troubleshooting.html');">Troubleshooting</a></td></tr>
  			<tr><td><a href="javascript:pageLink('chromaseqList.html');">Chromaseq E-mail List</a></td></tr>
  			<tr><td><a href="javascript:pageLink('limitations.html');">Limitations</a></td></tr>
  			<tr><td><a href="javascript:pageLink('otherPrograms.html');">Chromaseq &amp; Other Programs</a></td></tr>
  			<tr><td><a href="javascript:pageLink('future.html');">Future Directions</a></td></tr>
	</table>
  </div>
</div>

<div id="TechnicalPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Technical Details ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('sourceCode.html');">Source Code</a></td></tr>
  			<tr><td><a href="javascript:pageLink('technicalFAQ.html');">Technical FAQ</a></td></tr>
		</table>
  </div>
</div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('knownIssues.html')">Known Issues</a>
  </div>
  <div id="NonCollapsiblePanel">
    <a href="javascript:pageLink('versions.html')">Versions</a>
  </div>

<div id="CreditPanel" class="CollapsiblePanel">
  <div class="CollapsiblePanelTab" tabindex="0">Citation & Credit ˇ</div>
  <div class="CollapsiblePanelContent" style="display: block">
  		<table width="100%" border="0" cellpadding="1">
  			<tr><td><a href="javascript:pageLink('citation.html');">How to Cite</a></td></tr>
  			<tr><td><a href="javascript:pageLink('acknowledgements.html');">Acknowledgements</a></td></tr>
		</table>
  </div>
</div>


<h3>&nbsp;</h3>
  <!-- end #sidebar1 --></div>

  <div id="mainContent">
    
<!-- InstanceBeginEditable name="content" -->
  <p>In this tutorial,
      you will go from chromatograms to having fully processed sequences, with
      final base calls. Much of what is described here only needs to be done the first time you use Chromaseq.</p>
  <h2>Initial Setup</h2>
  <p>First, make sure that you have <a href="javascript:pageLink('installation.html');">installed Chromaseq, Phred, and Phrap</a>. </p>
  <p>Download the <a href="http://mesquiteproject.org/packages/chromaseq/examples/ChromaseqExamples.zip">ChromaseqExamples archive</a>,
    and unzip it. You will have a folder containing two subfolders and two files:</p>
  <ul>
    <li>two folders, each containing some example
            chromatograms</li>
    <li>a file called &quot;Primers.txt&quot; containing the <a href="javascript:pageLink('primers.html');">names and sequences
                of primers, their direction, and the gene fragment to which they correspond</a>,</li>
    <li>a file called &quot;SampleCodes.txt&quot; showing the correspondence between <a href="javascript:pageLink('sampleCode.html');">sample
                codes and sequence names</a></li>
    </ul>
  <p>You will also need to modify the phredpar.dat file that came with Phred,
    so that it can accommodate the chromatograms provided in the example. Open
    up the phredpar.dat file in a simple text editor, and find the line</p>
  <p class="monospaced"><font face="Courier New, Courier, mono">  &nbsp;&nbsp;&quot;BeckmanCEQ&quot; terminator
    big-dye	Beckman_CEQ_2000</font></p>
  <p>and immediately below this add the following two lines:</p>
  <p class="monospaced"><font face="Courier New, Courier, mono">&nbsp;&nbsp;#Entry below added to accommodate Chromaseq example files<br />
&nbsp;&nbsp;&quot;DT3730POP7{BDv3}.mob&quot; terminator big-dye ABI_373_377</font></p>
  <p>With that done, you are ready to process some chromatograms. We are going to first process the 14 chromatogram files that are in the folder &quot;Chromatograms&quot;.</p>
  <h2>Processing the Chromatograms  </h2>
  <h3>Telling Chromaseq how the chromatogram files are named</h3>
  <p>From the File&gt;Open Other submenu in Mesquite, choose &quot;Process Chromatograms Files in Directory&quot;. You will be first asked to choose the rules used to name chromatogram files.</p>
  <p align="center"><img src="images/namingRuleChoiceEmpty.png" width="318" height="138" /></p>
  <p>If this is the first time Chromaseq has been used on this computer, there will be no naming rules. Let's presume that is the case. To create a new naming rule, touch on the &quot;Edit Naming Rules&quot; button, and you should see a dialog box, that helps you manage your various naming rules, like this:</p>
  <p align="center"><img src="images/editNamingRulesEmpty.png" width="397" height="256" /></p>
  <p>We want to make a new naming rule, so press the &quot;New&quot; button, and give the rule the name &quot;ExamplesNamingRule&quot;</p>
  <p align="center"><img src="images/newNamingRuleEmpty.png" width="253" height="135" /></p>
  <p>You will then be presented with the editor in which you can specify the naming rule:</p>
  <p align="center"><img src="images/editingNamingRuleEmpty.png" width="321" height="296" /></p>
  <p>If you look into the &quot;Chromatograms&quot; folder, you will see that the file names look like this:</p>
  <p class="monospaced">	A01_A01DNA1120_B1490_411940.ab1</p>
  <p>The first 7 characters (&quot;A01_A01&quot;) specify the position in the 96-well sequencing plate; we don't care about this; the last bit, &quot;411940&quot;, is also not important. The only thing we care about in this is the sample code (which will allow us to figure out what to call the sequences that are created), and the primer name. In this file, the sample code is &quot;1120&quot;, and the primer name is &quot;B1490&quot;. We need to tell Chromaseq how to find these two items. </p>
  <p>As with all the files in that Chromatograms folder, the sample code precedes the primer name. The sample code is bounded by &quot;DNA&quot; and &quot;_&quot;. The primer name has a &quot;_&quot; in front of it, and a &quot;_&quot; following it. With those specified in the File Naming Rule dialog box, the naming rule is build. Not surprisingly, the default naming rule matches the file naming format used in these example files. </p>
  <p>Press OK, and then Done, and then you will be back to the dialog box used to choose your naming rule. Make sure you choose the rule you just created:</p>
  <p align="center"><img src="images/examplesNamingRule.png" width="382" height="138" /></p>
  <p>and press OK.</p>
  <h3>Choosing the method for naming sequences</h3>
  <p>You will then be asked to choose a module to be used to determine the names to give your processed sequences. There are several choices, three of which require you to have a file that provides a translation between sample codes and sequence names. The last one &quot;Use Sample Code for Sequence Name&quot; simply uses the sample code as the sequence name.</p>
  <p align="center"><img src="images/sequenceNameSupplier.jpg" width="280" height="333" /></p>
  <p>Choose the &quot;Sequence Names from Tab-delimited Text File&quot; option, and press OK. You will then be asked for the location of the file containing the sequence names; use the Browse button to choose the &quot;SampleCodes.txt&quot; file in the ChromaseqExamples folder:</p>
  <p align="center"><img src="images/exampleSampleCodeFileLocation.png" width="384" height="127" /></p>
  <p>This file is already set up to contain the sample codes (e.g., &quot;1120&quot;) and the name that should be applied to sequences from that sample (e.g., &quot;Bembidion perspicuum AZ 1120&quot;).</p>
  <h3>Choosing the file containing information about primers</h3>
    <p>Next, you will be asked to specify the location of the file that contains information about primers. Choose the &quot;Primers.txt&quot; file in the ChromaseqExamples directory:	</p>
    <p align="center"><img src="images/examplePrimerFileLocation.png" width="393" height="127" /></p>
    <h3>And now to make base calls and contigs...</h3>
  <p>We are getting close to being ready to do the processing. You should now see a dialog box that allows you to set things up for Phred and Phrap.</p>
  <p align="center"><img src="images/phredPhrapDialogTutorial.png" width="391" height="501" /></p>
  <p>We won't worry about most of this; the only thing we must do is to tell Chromaseq where the key helper programs it uses, including Phred and Phrap, are located. </p>
  <p>Press on the
      &quot;Phred Phrap Locations &amp; Options...&quot; button, and the dialog will appear
    that will
      let
      you tell Mesquite
      where
      the programs are located, and where the Phred parameter file is:</p>
  <p align="center"><img src="images/phredPhrapLocations.png" width="524" height="219" /></p>
  <p>Press the &quot;Browse...&quot; button beside the &quot;Phred, Phrap, &amp; Phd2Fasta&quot; to browse
    for the folder on your hard disk containing the three programs Chromaseq
    uses
    (Phred,
    Phrap,
    and
    Phd2Fasta),
    all of which must be in the same directory. Do the same for the Phred parameter
    file (&quot;phredpar.dat&quot;). You should only have to do this once on your computer,
    unless you move these files to other locations. </p>
  <p>Press OK on this &quot;Phred Phrap Locations &amp; Options&quot; dialog, and you will go back to the &quot;Phred/Phrap/Chromaseq Options&quot; dialog box. Press Process, and Chromaseq will then ask you for the directory that contains the chromatogram files to be processed. Find the chromatograms folder:</p>
  <p align="center"><img src="images/selectDirectory.png" width="423" height="346" /></p>
  <p>And press Choose, or OK, as appropriate.</p>
  <p>At this point, your computer's terminal shell should start. On a Mac, this
    will typically be the Terminal application; on Windows, it will be the Cygwin
    shell if that is what you installed. Phred and Phrap will then start processing
    the files, and the shell will be filled with the output of these programs.
    If you switch back to Mesquite, you will notice in the log information about
    what is happening. For example, here is the full log for processing the Chromatograms folder:  </p>
  <p>&nbsp;</p>
  <pre class="monospacedSmall">Phred Phrap processing of chromatograms as scripted by Mesquite
 Processing directory: /ChromaseqExamples/Chromatograms</pre>
  <pre class="monospacedSmall">Using names and codes file: /ChromaseqExamples/SampleCodes.txt</pre>
  <pre class="monospacedSmall">Using primers file: /ChromaseqExamples/Primers.txt</pre>
  <pre class="monospacedSmall">Parameters: Phred Options:  -id . -pd . -d 
     Phrap Options:  -new_ace -qual_show 20 -vector_bound 0 
     Chromaseq Options: 
     Quality threshold for lower case: 49
     Convert multiple-peaks sites to ambiguity codes: true
     Minimum secondary peak fraction for ambiguity: 0.3
     Trim low quality regions from ends: true
     Quality threshold for trim: 20
     Trim window length: 10
     Trim window threshold: 5 </pre>
  <pre class="monospacedSmall">Preparing Bemb perspicuum AZ 1120  (COI)
     Preparing Bemb louisella NS 1250  (COI)
     Preparing Bemb louisella NS 1250  (28S)
     Preparing Bemb perspicuum CA 2320  (28S)
     Preparing Bemb paraenulum VA 1856  (COI)
     Preparing Bemb perspicuum AZ 1120  (28S)
     Preparing Bemb paraenulum VA 1856  (28S)
     Number of files prepared: 7</pre>
  <pre class="monospacedSmall"> A01_A01DNA1120_B1490_411940.ab1  renamed to  DNA1120.b.B1490.ab1
     A01_A01DNA1120_Bcoi2R_412036.ab1  renamed to  DNA1120.g.Bcoi2R.ab1
     C07_C07DNA1250_B1490_208508.ab1  renamed to  DNA1250.b.B1490.ab1
     C07_C07DNA1250_B2232_208604.ab1  renamed to  DNA1250.g.B2232.ab1
     C12_C12DNA1250_D3aR_572645.ab1  renamed to  DNA1250.g.D3aR.ab1
     C12_C12DNA1250_NLF184_572549.ab1  renamed to  DNA1250.b.NLF184.ab1
     F06_F06DNA2320_D3aR_552091.ab1  renamed to  DNA2320.g.D3aR.ab1
     F06_F06DNA2320_NLF184_551995.ab1  renamed to  DNA2320.b.NLF184.ab1
     H05_H05DNA1856_B1490_237723.ab1  renamed to  DNA1856.b.B1490.ab1
     H05_H05DNA1856_B2232_237819.ab1  renamed to  DNA1856.g.B2232.ab1
     H07_H07DNA1120_D1_394061.ab1  renamed to  DNA1120.b.D1.ab1
     H07_H07DNA1120_D3_394157.ab1  renamed to  DNA1120.g.D3.ab1
     H09_H09DNA1856_D1_237563.ab1  renamed to  DNA1856.b.D1.ab1
     H09_H09DNA1856_D3_237659.ab1  renamed to  DNA1856.g.D3.ab1
  
Execution of Phred-Phrap command file begun.</pre>
<pre class="monospacedSmall">....1. Bemb perspicuum AZ 1120 (Phrap)
     2. Bemb louisella NS 1250 (Phrap)
     3. Bemb louisella NS 1250 (Phrap)
     4. Bemb perspicuum CA 2320 (Phrap)
     5. Bemb paraenulum VA 1856 (Phrap)
     6. Bemb perspicuum AZ 1120 (Phrap)
     7. Bemb paraenulum VA 1856 (Phrap)</pre>
  <pre class="monospacedSmall">Phred &amp; Phrap analyses completed.
     Post-Phrap processing begins.
     
     Processing ACE file: Bemb louisella NS 1250.ace
     C12_C12DNA1250_D3aR_572645.ab1 (rev)
     C12_C12DNA1250_NLF184_572549.ab1</pre>
  <pre class="monospacedSmall">Processing ACE file: Bemb paraenulum VA 1856.ace
     H09_H09DNA1856_D3_237659.ab1 (rev)
     H09_H09DNA1856_D1_237563.ab1</pre>
  <pre class="monospacedSmall">Processing ACE file: Bemb perspicuum AZ 1120.ace
     H07_H07DNA1120_D3_394157.ab1 (rev)
     H07_H07DNA1120_D1_394061.ab1</pre>
  <pre class="monospacedSmall">Processing ACE file: Bemb perspicuum CA 2320.ace
     F06_F06DNA2320_D3aR_552091.ab1 (rev)
     F06_F06DNA2320_NLF184_551995.ab1</pre>
  <pre class="monospacedSmall">Processing ACE file: Bemb louisella NS 1250.ace
     C07_C07DNA1250_B2232_208604.ab1 (rev)
     C07_C07DNA1250_B1490_208508.ab1</pre>
  <pre class="monospacedSmall">Processing ACE file: Bemb paraenulum VA 1856.ace
     H05_H05DNA1856_B2232_237819.ab1 (rev)
     H05_H05DNA1856_B1490_237723.ab1</pre>
  <pre class="monospacedSmall">Processing ACE file: Bemb perspicuum AZ 1120.ace
     A01_A01DNA1120_Bcoi2R_412036.ab1 (rev)
     A01_A01DNA1120_B1490_411940.ab1</pre>
  <pre class="monospacedSmall">========================
     Phred/Phrap processing report</pre>
  <pre class="monospacedSmall">Multi-read sequences obtained from the following genes and samples:
     (28S) -- Bembidion louisella NS 1250
     (28S) -- Bembidion paraenulum VA 1856
     (28S) -- Bembidion perspicuum AZ 1120
     (28S) -- Bembidion perspicuum CA 2320
     (COI) -- Bembidion louisella NS 1250
     (COI) -- Bembidion paraenulum VA 1856
     (COI) -- Bembidion perspicuum AZ 1120
     ========================</pre>
  <p>Your log should look like that above. </p>
  <p>Eventually, the processing will finish, and Mesquite will
    import the resulting contigs into a Mesquite project, with one matrix for
    each gene.</p>
  <p>In the example in this tutorial, the chromatograms come from two different
    genes, and so two matrices will appear in the project once the processing
    is done, one for 28S rDNA, and one for COI. The last matrix processed will be presented
    to you, by default in Birds Eye View, and with the bases colored according
    to the quality score assigned to them by Phred and Phrap. White is high quality,
    darker greens are moderate quality, and blue is poorest quality:</p>
  <p align="center"><img src="images/tutorialProcessed1.png" width="558" height="281" /></p>
  <p>If you want to see the matrix with the colors being the standard ones for nucleotides, choose Matrix&gt;Color Matrix Cells&gt;Character State, and it should look like this:</p>
  <p align="center"><img src="images/tutorialProcessByState.png" width="565" height="286" /></p>
  <p>We could now align the sequences, and do other things with them, but let's now worry about that now, instead let's look at the contigs and chromatograms in detail.  </p>
  <h2>Examining a contig and its chromatograms</h2>
  <p>To see the contig for the COI sequence of Bembidion louisella NS 1250 and its chromatograms,
    touch the view chromatograms tool (<img src="images/chromTool.jpg" width="30" height="29" align="absmiddle" />)
    on &quot;Bembidion louisella NS 1250&quot;. A new window will open, showing you the chromatograms
    and the contig:</p>
  <p align="center"><img src="images/tutorialProcessedViewer.png" width="624" height="393" /></p>
  <p>You may see a different part of the sequence than is shown in the above
    image; Chromaseq centers the view on the base upon which you clicked in the
    matrix.</p>
  <p align="center"><img src="images/tutorialFirstChromView.gif" width="657" height="278" /></p>
  <h3>Magnifying peaks</h3>
  <p>To the right of the reads is a close-up window, which shows a magnified view of a few peaks in one of the reads; the section that is being magnified is indicated in the reads by a blue-shaded box:  </p>
  <p align="center"><img src="images/tutorialProcessedViewer.png" width="624" height="393" /></p>
  <p>By default, the bases that are magnified are those in the center of the chromatogram window, and the read that is shown is the <em>source</em> read. The source read is the primary read that Phrap uses in making the base call. Source read for any particular site is that read with a short, brown bar under the peak. As you scroll through the contig, the close-up window will continue to track the source read at the center site.</p>
  <p>To see a different region magnified, you can hold down the Option key, and move the mouse over the region you would like magnified.</p>
  <h3>Navigating through a contig</h3>
  <p>There are several mechanisms to navigate through a contig. The simplest is to use the scroll bar at the bottom of the window. You can also scroll through the overview window, and then touch on the point of the overview you wish to see in the main part of the chromatogram viewer. </p>
  <p>There are also several keyboard commands that can move you quickly through the chromatogram. If you touch the left or right arrow keys, by default Chromaseq will jump to the previous or next &quot;problematic&quot; base. (Note: you may need to touch the mouse on one of the chromatograms before this will work; this is a bug that we hope to fix soon.) Problematic bases include, by default, very low quality bases, sites with a conflict between the source read and a higher-quality read, and so on. You can adjust the sorts of problems you wish to jump between using the Chromatograms&gt;Arrow Keys submenu.</p>
  <p>If you touch the up arrow key, you will jump to the start of the contig; touching the down arrow key will jump to the end of the contig.</p>
  <h3>Making base calls</h3>
  <p>Now touch on the 'Character Matrix &quot;28S (from Phred/Phrap)&quot;' tab at the start of the Mesquite window. You will then see the matrix of 4 processed sequences of 28S ribosomal DNA. Using the chromatogram tool, touch on the start of the sequence for Bembidion perspicuum 2320. You should see, near the start of the contig, the following section:</p>
  <p align="center"><img src="images/tutorialBaseCall1.png" width="137" height="292" /></p>
  <p>Note that the base call by Phred/Phrap/Mesquite is gcgaMaggg, but it looks as if the &quot;M&quot; really should be two separate bases, AC. Thus, the base calls should probably be gcgaacaggg. To change the M to AC, you can do the following.</p>
  <p>First, you will need to create a new base here. Use the Insert Base tool (<img src="images/insertBaseTool.png" width="30" height="31" />), and touch it between the M and the a on the bottom row:</p>
  <p align="center"><img src="images/tutorialBaseCall2.png" width="137" height="192" /></p>
  <p>This will open up a new base (left, below), which you can then change by typing the letter for the base you wish to set it to, for example, &quot;c&quot; (right, below)</p>
  <p align="center"><img src="images/tutorialBaseCall3.png" width="137" height="192" />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img src="images/tutorialBaseCall4.png" width="137" height="192" /></p>
  <p>You can then touch on the M to select it, and then type an &quot;a&quot; to change it to an A (left, below). If you click somewhere else in the window to select the base you were editing, you will more clearly see that this has revised the sequence to gcgaacaggg (right, below). Bases you have manually edited are highlighted in yellow.</p>
  <p align="center"><img src="images/tutorialBaseCall5.png" width="137" height="192" />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img src="images/tutorialBaseCall6.png" width="137" height="192" /></p>
  <p>If you look in the data matrix itself, and turn off Bird's Eye View (by unchecking it under Matrix&gt;Display&gt;Bird's Eye View), you will see the new base calls highlighted in yellow:</p>
  <p align="center"><img src="images/tutorialBaseCall7.png" width="434" height="22" /></p>
  <h3>Visualizing and removing primer sites</h3>
  <p>To see where primer sites are in your sequence, you will need to ask Chromaseq to search for and show them. To do this, choose Chromatograms&gt;Primers&gt;Show Primer Matches. You will be asked for the location of the primer text file (for details, see the <a href="primers.html">Primer &amp; Gene Information</a> page), and either browse for the file or type in its path.</p>
  <p align="center"><img src="images/tutorialPrimerFile.png" width="497" height="145" /></p>
  <p>If you do this for the Bemb perspicuum AZ 1120 sequence in the COI matrix, and look at the start of the contig, you will see the B1490 primer drawn in gray letters over the peaks near the start of the sequence; you can see the match between the primer sequence and the bases in that read.</p>
  <p align="center"><img src="images/tutorialShowPrimer2.png" width="384" height="107" /></p>
  <p>(Note: if you want Chromaseq to show the primers for the next chromatogram window you open and subsequent windows, you will need to choose Chromatogram&gt;Save Current Settings as Defaults after you choose Chromatograms&gt;Primers&gt;Show Primer Matches.)</p>
  <p>To remove this region from the sequence in the matrix, first select the primer region in the chromatogram viewer. Do this by touching on the start of the primer, then, while holding down the Shift key, touch on the last base of the primer:  </p>
  <p>&nbsp;</p>
  <p align="center"><img src="images/tutorialToGaps1.png" width="388" height="319" /></p>
  <p>&nbsp;</p>
  <p>With those bases selected, if you choose Chromatograms&gt;Convert Selected to Gaps, those sites will be converted to gaps in both the chromatogram viewer and in the data matrix:  </p>
  <p>&nbsp;</p>
  <p align="center"><img src="images/tutorialToGaps2.png" width="388" height="319" /></p>
  <p>&nbsp;</p>
  <p>&nbsp;</p>
  <p>&nbsp;</p>
  <p>&nbsp;</p>
  <p>&nbsp;</p>
<!-- InstanceEndEditable -->

    
    
    </div>
	<!-- This clearing element should immediately follow the #mainContent div in order to force the #container div to contain all child floats --><br class="clearfloat" />
  <div id="footer">
    <p> ©2018 David R. Maddison &amp; Wayne P. Maddison</p>
  <!-- end #footer --></div>
<!-- end #container --></div>

<script type="text/javascript">
<!--


var GettingStartedPanel = null;
var ChromatogramViewerPanel = null;
var ProcessingChromatoPanel = null;
var PreparingDataPanel = null;
var CreditPanel = null;
var HelpPanel = null;
var GenBankPanel = null;
var OtherFeaturesPanel = null;
var TechnicalPanel = null;

if (qsParm['ProcessingChromatoPanel']=='open')
 	ProcessingChromatoPanel = new Spry.Widget.CollapsiblePanel("ProcessingChromatoPanel", {contentIsOpen: true});
else
 	ProcessingChromatoPanel = new Spry.Widget.CollapsiblePanel("ProcessingChromatoPanel", {contentIsOpen: false});
	
if (qsParm['GettingStartedPanel']=='open')
 	GettingStartedPanel = new Spry.Widget.CollapsiblePanel("GettingStartedPanel", {contentIsOpen: true});
else
 	GettingStartedPanel = new Spry.Widget.CollapsiblePanel("GettingStartedPanel", {contentIsOpen: false});
	
if (qsParm['ChromatogramViewerPanel']=='open')
 	ChromatogramViewerPanel = new Spry.Widget.CollapsiblePanel("ChromatogramViewerPanel", {contentIsOpen: true});
else
 	ChromatogramViewerPanel = new Spry.Widget.CollapsiblePanel("ChromatogramViewerPanel", {contentIsOpen: false});

if (qsParm['PreparingDataPanel']=='open')
 	PreparingDataPanel = new Spry.Widget.CollapsiblePanel("PreparingDataPanel", {contentIsOpen: true});
else
 	PreparingDataPanel = new Spry.Widget.CollapsiblePanel("PreparingDataPanel", {contentIsOpen: false});

if (qsParm['CreditPanel']=='open')
 	CreditPanel = new Spry.Widget.CollapsiblePanel("CreditPanel", {contentIsOpen: true});
else
 	CreditPanel = new Spry.Widget.CollapsiblePanel("CreditPanel", {contentIsOpen: false});

if (qsParm['HelpPanel']=='open')
 	HelpPanel = new Spry.Widget.CollapsiblePanel("HelpPanel", {contentIsOpen: true});
else
 	HelpPanel = new Spry.Widget.CollapsiblePanel("HelpPanel", {contentIsOpen: false});

if (qsParm['GenBankPanel']=='open')
 	GenBankPanel = new Spry.Widget.CollapsiblePanel("GenBankPanel", {contentIsOpen: true});
else
 	GenBankPanel = new Spry.Widget.CollapsiblePanel("GenBankPanel", {contentIsOpen: false});
    
if (qsParm['OtherFeaturesPanel']=='open')
 	OtherFeaturesPanel = new Spry.Widget.CollapsiblePanel("OtherFeaturesPanel", {contentIsOpen: true});
else
 	OtherFeaturesPanel = new Spry.Widget.CollapsiblePanel("OtherFeaturesPanel", {contentIsOpen: false});
    
if (qsParm['TechnicalPanel']=='open')
 	TechnicalPanel = new Spry.Widget.CollapsiblePanel("TechnicalPanel", {contentIsOpen: true});
else
 	TechnicalPanel = new Spry.Widget.CollapsiblePanel("TechnicalPanel", {contentIsOpen: false});



//-->
</script>
</body>
<!-- InstanceEnd --></html>
